lib/src/room.dart

```diff
@@ -1150,21 +1150,29 @@ class Room {
   /// Call the Matrix API to join this room if the user is not already a member.
   /// If this room is intended to be a direct chat, the direct chat flag will
   /// automatically be set.
-  Future<void> join({bool leaveIfNotFound = true}) async {
+  Future<void> join({
+    /// In case of the room is not found on the server, the client leaves the
+    /// room and rethrows the exception.
+    bool leaveIfNotFound = true,
+  }) async {
+    final dmId = directChatMatrixID;
     try {
       // If this is a DM, mark it as a DM first, because otherwise the current member
       // event might be the join event already and there is also a race condition there for SDK users.
-      final dmId = directChatMatrixID;
-      if (dmId != null) {
-        await addToDirectChat(dmId);
-      }
+      if (dmId != null) await addToDirectChat(dmId);
 
       // now join
       await client.joinRoomById(id);
     } on MatrixException catch (exception) {
+      if (dmId != null) await removeFromDirectChat();
```

Comment:
Can we put this in the leave room call?

Reply:
The spec doesn't say so, but it is intuitive to update the m.direct list if you leave the room right? Or does this have any edge cases?

Reply:
We had this before and removed by a reason that in the archive we still want to see that this was once a DM room with this user.

Reply:
ah right! thanks